scylla_util.py: on sysconfig_parser, don't use double quote when it's possible

syuu1228 opened the pull request:
It seems like distribution original sysconfig files does not use double
quote to set the parameter when the value does not contain space.
Adding function to detect spaces in the value, don't usedouble quote
when it not detected.

Fixes #9149

penberg commented on dist/common/scripts/scylla_util.py:
```diff
@@ -1008,11 +1008,14 @@ def __load(self):
         self._cfg.readfp(f)
 
+    def __fmt(self, val):
+        return '{}="{}"' if ' ' in val else '{}={}'
```
Let's check for other whitespace characters too like tabs for extra safety?

```python
need_quotes = any([ch.isspace() for ch in val])
return '{}="{}"' if need_quotes else '{}={}'
```

syuu1228 replied:
@penberg Applied `need_quotes` to new version.

avikivity commented on dist/common/scripts/scylla_util.py:
```diff
@@ -1008,11 +1008,15 @@ def __load(self):
+    def __fmt(self, val):
+        need_quotes = any([ch.isspace() for ch in val])
+        return '{}="{}"' if need_quotes else '{}={}'
+
```
Better to use shlex.quote, it handles more cases:

```
>>> import shlex
>>> shlex.quote('x')
'x'
>>> shlex.quote('x"')
'\'x"\''
>>> shlex.quote('x y')
"'x y'"
>>> 
```

syuu1228 replied:
@avikivity it is probably correct that shlex.quote can handle more cases to escape, but what we want to implement on #9149 is, aligning output to distribution default style of sysconfig file.
Which is
 - don't use quotes when parameter does not include space
 - use double quotes when parameter does include space
Something like this:
```
SELINUX=enforcing
```
```
CPUPOWER_START_OPTS="frequency-set -g performance"
```

However, shlex.quote seems like always uses single quote, and don't able to customize it, so it doesn't fit for #9149.

syuu1228 commented:
ping

vladzcloudius reviewed (changes requested):
NACK

Based on @ebenzecri report (https://github.com/scylladb/scylladb/issues/9149#issuecomment-988336043) double quote should always be used.

vladzcloudius commented:
Oh, I was confused then. Let me re-review it then.

vladzcloudius commented on dist/common/scripts/scylla_util.py:
```diff
@@ -1008,11 +1008,15 @@ def __load(self):
 
     def __add(self, key, val):
-        self._data += '{}="{}"\n'.format(key, self.__escape(val))
+        self._data += self.__fmt(val).format(key, self.__escape(val)) + '\n'
```
This is weird: why do you both put quotes via `__fmt(...)` but also escape quotes via `__escape(val)`. What is the use case for this?

vladzcloudius replied:
@syuu1228 ping

syuu1228 replied:
> This is weird: why do you both put quotes via `__fmt(...)` but also escape quotes in the value itself via `__escape(val)`. What is the use case for this?

__escape(val) is just for when val contains double quote in middle of the value , something like `s3.set('ESCAPE1', 'aaaa"bbbb')` I wrote test code below.

> What if the value that arrives is `"disable"`? Don't we want to remove quotes in that case?

It will drop by strip('"') at get(), since strip() function removes both head of the string and tail of the string:
```
    def get(self, key):
        return self._cfg.get('global', key).strip('"')
```
https://github.com/scylladb/scylladb/blob/master/dist/common/scripts/scylla_util.py#L443

vladzcloudius replied:
I see. No more questions. Thanks. ;)

syuu1228 commented:
Rewrited the code since `self.__fmt(val).format(key, value)` was looks strange code, so I dropped it and replaced by `__format_line(self, key, val)` to return correctly formatted "{key}={val}" string.

Also I found that even we have __escape() for escaping " middle of the value to writing sysconfig file, we didn't unescape for reading from sysconfig file.
So adding __unescape() and call it on get().
(see: https://github.com/scylladb/scylladb/pull/9153/commits/cd6030d5df1f25b71c23bb6508fcb503acec2a88)

syuu1228 commented:
Verified not quoted value and quote value works as expected, on following test code.
Also tested escaping double quote worked as expected.

I will add the test code into #10809.

```
import sys
sys.path.append('./dist/common/scripts')
from scylla_util import sysconfig_parser

s1 = sysconfig_parser('cfg1')
s1.set('SELINUX', 'disabled')
s1.commit()
with open('cfg1') as f:
    s1str = f.read()
assert s1str == 'SELINUX=disabled\n'

s1 = sysconfig_parser('cfg1')
assert s1.get('SELINUX') == 'disabled'

s2 = sysconfig_parser('cfg2')
s2.set('SCYLLA_CONF', '/etc/scylla')
s2.set('SCYLLA_ARGS', '--log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix')
s2.commit()
with open('cfg2') as f:
    s2str = f.read()
assert s2str == 'SCYLLA_CONF=/etc/scylla\nSCYLLA_ARGS="--log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix"\n'

s2 = sysconfig_parser('cfg2')
assert s2.get('SCYLLA_CONF') == '/etc/scylla'
assert s2.get('SCYLLA_ARGS') == '--log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix'

s3 = sysconfig_parser('cfg3')
s3.set('ESCAPE1', 'aaaa"bbbb')
s3.set('ESCAPE2', 'aaaa bbbb"cccc')
s3.commit()
with open('cfg3') as f:
    s3str = f.read()
assert s3str == 'ESCAPE1=aaaa\\"bbbb\nESCAPE2="aaaa bbbb\\"cccc"\n'

s3 = sysconfig_parser('cfg3')
assert s3.get('ESCAPE1') == 'aaaa"bbbb'
assert s3.get('ESCAPE2') == 'aaaa bbbb"cccc'
```

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2782/

vladzcloudius reviewed (approved):
LGTM

mykaul commented:
@roydahan - please note that what's hidden behind this fix is actually allow selinux.

It's unclear to me what our selinux policy is, but let's verify that it's not disabled.

roydahan commented:
IIRC we recommend to have it as disabled and I think that the issue here was that it wasn’t disabled.

vladzcloudius commented:
> @roydahan - please note that what's hidden behind this fix is actually allow selinux to be set to other than 'disabled'.
> 
> It's unclear to me what our selinux policy is, but let's verify that it's not disabled.

It's actually quite the opposite, @mykaul.
This patch restores the original scylla behavior - which was disabling `selinux`.

mykaul commented:
> > @roydahan - please note that what's hidden behind this fix is actually allow selinux to be set to other than 'disabled'.
> > It's unclear to me what our selinux policy is, but let's verify that it's not disabled.
> 
> It's actually quite the opposite, @mykaul. This patch restores the original scylla behavior - which was (and is now) disabling `selinux`.

Understood - which I don't see the value in, unless there's a very very convincing argument for it.

avikivity commented:
It was performance. I don't remember the results, it wasn't huge, but it was significant.
